File: C/render_mini.cpp


//calc
#include <cstdlib>
#include <cmath>
#include <algorithm> 
#include <time.h>
#include <vector>
#include <iostream>
#include <array>
#include <time.h>
#include <limits>
#include <cstddef>
#include <chrono>
#include <ctime>
#include <ratio>


//draw
#include <gtk/gtk.h>
#include <cairo.h>
#include <cairo-pdf.h>
#include <cairo-ps.h>
#include <cairo-svg.h>
#include <gdk/gdkkeysyms.h>
#include <X11/Xlib.h>
#include <thread>
#include <gdk/gdkkeysyms.h>

using namespace std;

const int viewWidth = 300;
const int viewHeight = 200;
const int viewScaleX = 3;
const int viewScaleY = 3;
bool active = true;
int scene = 1;

GtkWidget *canvas = NULL;
std::thread *draw_thread = NULL;


static void draw_gtk(GtkWidget *, GdkEventExpose *, gpointer);
bool keyDown[65536]={};
class Gg2 {
private:

    bool disable_ani = false; //set true to disable animation
    double speed = 0.12;
    double leftm = 0;
    double rightm = 0;
    double zz = 0;
    double wlight[3] = {0, 0, 0};
    bool wlightstate[3] = {false, false, false};
    bool wlightprgrs[3] = {false, false, false};
    double maxinverval = 3.5;
    double inverval = maxinverval;
    double invervalt[3] = {maxinverval, maxinverval, 1};
    bool invert = false;

    bool currcolor[3] = {true, true, false};

    double lastatime = 1000;

    int mat_blue = 0x0000ff;
    int mat_yellow = 0xffff00;


    double frameRateSmoothing = 1.0;


    double maxViewDist = 100;
    double primaryRayStepThreshold = 0.01;
    double noIntersection = -1.0;


    class Vec {
    public:

        double x;
        double y;
        double z;

        Vec(double x, double y, double z) {
            this->x = x;
            this->y = y;
            this->z = z;
        }

        double length() {
            return std::sqrt(x * x + y * y + z * z);
        }

        double dotProduct(Vec v) {
            return x * v.x + y * v.y + z * v.z;
        }

        Vec add(Vec v) {
            return Vec(x + v.x, y + v.y, z + v.z);
        }

        Vec subtract(Vec v) {
            return Vec(x - v.x, y - v.y, z - v.z);
        }

        Vec multiply(double a) {
            return Vec(x * a, y * a, z * a);
        }

        Vec normalise() {
            return this->multiply(1.0 / length());
        }

        Vec crossProduct(Vec v) {
            return Vec(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x);
        }
    };

    static void closethis() {
        active = false;
        std::this_thread::sleep_for(std::chrono::seconds(1));
        gtk_main_quit();

    }

    static gboolean processKeyEvent(GtkWidget *widget, GdkEventKey *event, gpointer user_data) {

        switch (event->keyval) {
            case GDK_1:scene=1;keyDown[GDK_1] = true;
                break;
            case GDK_2:scene=2;keyDown[GDK_2] = true;
                break;
            case GDK_s:keyDown[GDK_s] = true;
                break;
            case GDK_w:keyDown[GDK_w] = true;
                break;
            case GDK_a:keyDown[GDK_a] = true;
                break;
            case GDK_d:keyDown[GDK_d] = true;
                break;
            case GDK_q:keyDown[GDK_q] = true;
                break;
            case GDK_e:keyDown[GDK_e] = true;
                break;
            default:
                return FALSE;
        }
        return FALSE;
    }

    static void gtk_draw() {
        XInitThreads();
        gtk_init(NULL, NULL);
        GtkWidget *mainwin = NULL;

        mainwin = gtk_window_new(GTK_WINDOW_TOPLEVEL);
        canvas = gtk_drawing_area_new();
        gtk_widget_set_size_request(canvas, viewWidth*viewScaleX, viewHeight * viewScaleY);

        gtk_container_add(GTK_CONTAINER(mainwin), canvas);

        g_signal_connect(mainwin, "destroy", G_CALLBACK(closethis), NULL);

        g_signal_new("paint-this",
                G_TYPE_OBJECT, G_SIGNAL_RUN_FIRST,
                0, NULL, NULL,
                g_cclosure_marshal_VOID__POINTER,
                G_TYPE_NONE, 1, G_TYPE_POINTER);

        g_signal_connect(G_OBJECT(canvas), "paint-this", G_CALLBACK(draw_gtk), NULL);
        g_signal_connect(G_OBJECT(mainwin), "key_press_event", G_CALLBACK(processKeyEvent), NULL);

        gtk_widget_show_all(mainwin);
        gtk_main();
    }

    double getDistToSphere(Vec pt, double radius) {
        return pt.length() - radius;
    }

    double getDistToBox(Vec pt, Vec halfSize) {
        Vec tmp = Vec(std::max((double) 0, std::abs(pt.x) - halfSize.x),
                std::max((double) 0, std::abs(pt.y) - halfSize.y),
                std::max((double) 0, std::abs(pt.z) - halfSize.z));
        return tmp.length();
    }

    Vec tilePrimitive(Vec pt, double tileSize) {

        double halfTileSize = tileSize / 2.0;
        return Vec(fmod(pt.x, tileSize) - halfTileSize,
                fmod(pt.y, tileSize) - halfTileSize,
                fmod(pt.z, tileSize) - halfTileSize);
    }

    Vec getSphereNormal(Vec pt) {
        return pt.normalise();
    }

    double worldGetDist(Vec pt) {

        double tileSize = 3.0;
        pt = tilePrimitive(pt, tileSize);

        switch (scene) {
            case 1:
                return getDistToSphere(pt, 1);
            case 2:
                return getDistToBox(pt, Vec(0.5, 0.5, 0.5));
            default:
                return getDistToSphere(pt, 1);
        }
    }

    Vec worldGetNormal(Vec pt) {

        double tileSize = 3.0;
        pt = tilePrimitive(pt, tileSize);
        return getSphereNormal(pt);
    }

    double worldIntersect(Vec rayStart, Vec rayDir, double maxDist, double rayStepThreshold) {

        Vec pt = rayStart;
        double rayLen = 0.0;

        double dist;

        while ((dist = worldGetDist(pt)) > rayStepThreshold) {
            rayLen += dist;
            if (rayLen > maxDist) {
                rayLen = noIntersection;
                break;
            }

            pt = pt.add(rayDir.multiply(dist));
        }

        return rayLen;
    }

    void putpixel(int x, int y, int color) {
        if (x < 0 || x >= viewWidth || y < 0 || y >= viewHeight) {
            return;
        }

        pixels[y * viewWidth + x] = color;
    }

public:
    
    int pixels [viewWidth * viewHeight];
    long frame_timetocalc = 0;
    long min_timetocalc = 1000000000;
    long max_timetocalc = 0;
    int fps = 0;

    void run() {

        int numFrames = 0;
        fps = 0;
        std::chrono::high_resolution_clock::time_point lastFpsTime;
        while (active) {
            clock_t begin = clock();
            if (keyDown[GDK_w]) {
                keyDown[GDK_w] = false;
                speed = (speed < 0.3) ? speed + 0.05 : 0.35;

            }

            if (keyDown[GDK_s]) {
                keyDown[GDK_s] = false;
                speed = (speed > 0.06) ? speed - 0.05 : 0;

            }
            lastatime += speed;

            if (keyDown[GDK_a]) {
                keyDown[GDK_a] = false;
                if (rightm == 0) {
                    leftm = (leftm < 0.9) ? leftm + 0.08 : 0.91;
                } else {
                    rightm = ((rightm - 0.08) > 0) ? rightm - 0.08 : 0;
                }
            }

            if (keyDown[GDK_d]) {
                keyDown[GDK_d] = false;
                if (leftm == 0) {
                    rightm = (rightm < 0.9) ? rightm + 0.08 : 0.91;
                } else {
                    leftm = ((leftm - 0.08) > 0) ? leftm - 0.08 : 0;
                }
            }

            if (keyDown[GDK_q]) {
                keyDown[GDK_q] = false;
                zz = (zz < 0.0055) ? zz + 0.0005 : 0.006;
            }

            if (keyDown[GDK_e]) {
                keyDown[GDK_e] = false;
                zz = ((zz - 0.0005) > -0.0055) ? zz -= 0.0005 : -0.006;
            }

            Vec moveDir(1.5, zz, 0.8);
            Vec cameraPos = moveDir.multiply(lastatime);
            Vec cameraLookAt = cameraPos.add(moveDir).add(Vec(0 + rightm, 0.8 - zz * 500, 0.5 + leftm));
            Vec cameraUp(0 + rightm / 3, 1.0, 0 + leftm / 3);

            Vec cameraForward = cameraLookAt.subtract(cameraPos).normalise();
            Vec cameraRight = cameraForward.crossProduct(cameraUp);
            cameraUp = cameraRight.crossProduct(cameraForward);

            if (!(wlightprgrs[0] || wlightprgrs[1] || wlightprgrs[2])) {
                wlightprgrs[0] = true;
            }
            for (int ij = 0; ij < 3; ij++) {
                if (wlightprgrs[ij]) {
                    if (wlightstate[ij]) {
                        wlight[ij] -= 0.05;
                        if (wlight[ij] < 0.1) {
                            wlight[ij] = 0.05;
                            wlightstate[ij] = false;
                            wlightprgrs[ij] = false;
                            if (ij != 2) {
                                wlightprgrs[ij + 1] = true;
                            }
                        }
                    } else {
                        wlight[ij] += 0.05;
                        if (wlight[ij] > 1) {
                            wlight[ij] = 1;
                            wlightstate[ij] = true;
                            wlightprgrs[ij] = false;
                            if (ij != 2) {
                                wlightprgrs[ij + 1] = true;
                            }
                        }
                    }
                    break;
                }
            }
            if (!disable_ani) {
                if (inverval + 0.05 > maxinverval) {
                    inverval = maxinverval;
                    invert = true;
                }
                if (invert) {
                    inverval -= 0.05;
                    if (currcolor[0] && currcolor[1]) {
                        invervalt[0] -= 0.05;
                        invervalt[2] += 0.05;
                    } else if (currcolor[1] && currcolor[2]) {
                        invervalt[1] -= 0.05;
                        invervalt[0] += 0.05;
                    } else if (currcolor[2] && currcolor[0]) {
                        invervalt[2] -= 0.05;
                        invervalt[1] += 0.05;
                    }
                } else {
                    inverval += 0.05;
                    if (currcolor[0] && currcolor[1]) {
                        invervalt[0] += 0.05;
                        invervalt[1] -= 0.05;
                    } else if (currcolor[1] && currcolor[2]) {
                        invervalt[1] += 0.05;
                        invervalt[2] -= 0.05;
                    } else if (currcolor[2] && currcolor[0]) {
                        invervalt[2] += 0.05;
                        invervalt[0] -= 0.05;
                    }
                }
                if (inverval - 0.04 > 1) {

                } else {
                    invert = false;
                    inverval = 1;

                    if (currcolor[0] && currcolor[1]) {
                        inverval = maxinverval;
                        currcolor[0] = false;
                        currcolor[2] = true;
                    } else if (currcolor[1] && currcolor[2]) {
                        inverval = maxinverval;
                        currcolor[1] = false;
                        currcolor[0] = true;
                    } else if (currcolor[2] && currcolor[0]) {
                        inverval = maxinverval;
                        currcolor[2] = false;
                        currcolor[1] = true;
                    }
                }
            } else inverval = invervalt[0] = invervalt[1] = invervalt[2] = 1;

            for (int row = 0; row < viewHeight; row++) {
                for (int col = 0; col < viewWidth; col++) {

                    double xFrac = (double) (col - viewWidth * 0.5) / viewWidth;
                    double yFrac = (double) (row - viewHeight * 0.5) / viewWidth;
                    double zFrac = 0.5;

                    Vec rayStart = cameraPos;
                    Vec rayDir = cameraRight.multiply(xFrac).add(cameraUp.multiply(yFrac)).add(cameraForward.multiply(zFrac));
                    rayDir = rayDir.normalise();
                    double depth = worldIntersect(rayStart, rayDir, maxViewDist, primaryRayStepThreshold);

                    if (depth == noIntersection) {
                        if (scene == 2) {
                            putpixel(col, row, mat_yellow);
                        } else {
                            putpixel(col, row, mat_blue);
                        }
                    } else {

                        Vec tmp(-1 - wlight[1] * 2, 0.5 - wlight[1] * 2, -0.7 - wlight[2] * 2);
                        Vec dirToLight = tmp.normalise();
                        double shadowFactor = noIntersection;

                        Vec pt = rayStart.add(rayDir.multiply(depth));
                        Vec normal = worldGetNormal(pt);

                        if (shadowFactor == noIntersection) {
                            shadowFactor = 1.0;
                        } else {
                            shadowFactor = 1.0 - shadowFactor;
                        }

                        double normalisedIntensity = std::min(std::max(0.0, normal.dotProduct(dirToLight)) * shadowFactor, 1.0);
                        int intensity = (int) (normalisedIntensity * 255.0);

                        int intensityr = (int) (intensity * invervalt[0]);
                        int intensityg = (int) (intensity * invervalt[1]);
                        int intensityb = (int) (intensity * invervalt[2]);
                        if (scene == 2) {
                            unsigned char depthByte = 0;
                            double invMaxViewDist = 1.0 / maxViewDist;
                            if (depth != noIntersection) {
                                depthByte = (unsigned char) ((1.0 - depth * invMaxViewDist) * 255.0);
                                putpixel(col, row, ((int) (depthByte - intensity * 0.5) << 16) + ((int) (depthByte - intensity * 0.5) << 8) + (int) (depthByte));
                            }
                        } else {
                            putpixel(col, row, ((intensityr) << 16) + ((intensityg) << 8) + intensityb);

                        }
                    }
                }
            }

            clock_t end = clock();
            frame_timetocalc = ((double) (end - begin) / CLOCKS_PER_SEC)*1000;
            if (max_timetocalc < frame_timetocalc) {
                max_timetocalc = frame_timetocalc;
            }
            if (min_timetocalc > frame_timetocalc) {
                min_timetocalc = frame_timetocalc;
            }

            if (active)
                g_signal_emit_by_name(canvas, "paint-this");
            numFrames++;
            std::chrono::duration<double> delta = std::chrono::duration_cast<std::chrono::duration<double>> (std::chrono::high_resolution_clock::now() - lastFpsTime);
            if (delta.count() > frameRateSmoothing) {
                fps = (int) (numFrames / delta.count());
                numFrames = 0;
                lastFpsTime = std::chrono::high_resolution_clock::now();
            }

        }
    }

    void start() {
        draw_thread = new std::thread(gtk_draw);
        run();
        draw_thread->join();

    }

};

Gg2 *newgg;

static void draw_gtk(GtkWidget *widget, GdkEventExpose *eev, gpointer data) {
    cairo_t *image;
    image = gdk_cairo_create(widget->window);
    cairo_scale(image, viewScaleX, viewScaleY);
    cairo_move_to(image, 0, 0);
    for (int i = 0; i < viewWidth; i++) {
        for (int j = 0; j < viewHeight; j++) {
            cairo_move_to(image, i, j);
            cairo_set_source_rgb(image, (newgg->pixels[j * viewWidth + i] >> 16) / (float) 255, ((newgg->pixels[j * viewWidth + i] >> 8)-((newgg->pixels[j * viewWidth + i] >> 16) << 8)) / (float) 255, ((newgg->pixels[j * viewWidth + i])-((newgg->pixels[j * viewWidth + i] >> 8) << 8)) / (float) 255);
            cairo_line_to(image, i, j + 1);
            cairo_close_path(image);
            cairo_stroke(image);

        }
    }

    cairo_select_font_face(image, "Sans", CAIRO_FONT_SLANT_NORMAL,
            CAIRO_FONT_WEIGHT_BOLD);
    cairo_set_font_size(image, 12.0);
    cairo_move_to(image, 0, viewHeight);
    cairo_text_path(image, (string("fps:") + string(std::to_string(newgg->fps)) + string(" , timetocalc:") + string(std::to_string(newgg->frame_timetocalc)) + string(" ,  max:") + string(std::to_string(newgg->max_timetocalc)) + string(" , min:") + string(std::to_string(newgg->min_timetocalc))).c_str());
    cairo_set_source_rgb(image, 1, 0, 0);
    cairo_fill_preserve(image);

    cairo_destroy(image);



}

int main(int argc, char** argv) {
    newgg = new Gg2;
    newgg->start();
    cout << "max:" << newgg->max_timetocalc << " , min:" << newgg->min_timetocalc << std::endl;

};

File: js/render_mini.js
	const {sqrt, max, min, abs, round, floor} = Math;
	class Vec {

		constructor(x, y, z) {
			this.x = x
			this.y = y
			this.z = z
			this.length = sqrt(x * x + y * y + z * z)
		}
		dotProduct(v) {
            return this.x * v.x + this.y * v.y + this.z * v.z
        }

        add(v) {
            return new Vec(this.x + v.x, this.y + v.y, this.z + v.z)
        }

        subtract(v) {
            return new Vec(this.x - v.x, this.y - v.y, this.z - v.z)
        }

        multiply(a) {
            return new Vec(this.x * a, this.y * a, this.z * a)
        }

        normalise() {
            const a = (1.0 / this.length)
            return new Vec(this.x * a, this.y * a, this.z * a)
        }

        crossProduct(v) {
            return new Vec(this.y * v.z - this.z * v.y, this.z * v.x - this.x * v.z, this.x * v.y - this.y * v.x)
        }
	}



function Gg2() {

}

Gg2.key_1 = 49;
Gg2.key_2 = 50;
Gg2.key_Q = 113;
Gg2.key_E = 101;
Gg2.key_W = 119;
Gg2.key_A = 97;
Gg2.key_S = 115;
Gg2.key_D = 100;

Gg2.speed = 0.12;
Gg2.leftm = 0;
Gg2.rightm = 0;
Gg2.zz = 0;
Gg2.wlight = [0, 0, 0];
Gg2.wlightstate = [false, false, false];
Gg2.wlightprgrs = [false, false, false];
Gg2.maxinverval = 3.5;
Gg2.disable_ani = false; //set true to disable animation
Gg2.inverval = Gg2.maxinverval;
Gg2.invervalt = [Gg2.maxinverval, Gg2.maxinverval, 1];
Gg2.invert = false;

Gg2.currcolor = [true, true, false];

Gg2.lastatime = 1000;
Gg2.viewWidth = 300;
Gg2.viewHeight = 200;
Gg2.viewScaleX = 3;
Gg2.viewScaleY = 3;

Gg2.mat_blue = 0x0000ff;
Gg2.mat_yellow = 0xffff00;

Gg2.pixels = new Array(Gg2.viewWidth * Gg2.viewHeight);

Gg2.frameRateSmoothing = 1.0;
Gg2.milsPerSecond = 1000;


Gg2.maxViewDist = 100;
Gg2.primaryRayStepThreshold = 0.01;
Gg2.noIntersection = -1.0;

Gg2.scene = 1;

Gg2.keyDown = new Array(65536);

Gg2.frame_timetocalc = 0;
Gg2.min_timetocalc = 1000000000;
Gg2.max_timetocalc = 0;

Gg2.getDistToSphere = function (pt, radius) {
    return pt.length - radius;
}

Gg2.getDistToBox = function (pt, halfSize) {
    return new Vec(Math.max(0, Math.abs(pt.x) - halfSize.x),
            Math.max(0, Math.abs(pt.y) - halfSize.y),
            Math.max(0, Math.abs(pt.z) - halfSize.z)).length;
}

Gg2.tilePrimitive = function (pt, tileSize) {

    var halfTileSize = tileSize / 2.0;
    return new Vec(pt.x % tileSize - halfTileSize,
            pt.y % tileSize - halfTileSize,
            pt.z % tileSize - halfTileSize);
}

Gg2.getSphereNormal = function (pt) {
    return pt.normalise();
}

Gg2.worldGetDist = function (pt) {

    var tileSize = 3.0;
    pt = Gg2.tilePrimitive(pt, tileSize);

    switch (Gg2.scene) {
        case 1:
            return Gg2.getDistToSphere(pt, 1);
        case 2:
            return Gg2.getDistToBox(pt, new Vec(0.5, 0.5, 0.5));
        default:
            return Gg2.getDistToSphere(pt, 1);
    }
}

Gg2.worldGetNormal = function (pt) {

    var tileSize = 3.0;
    pt = Gg2.tilePrimitive(pt, tileSize);
    return Gg2.getSphereNormal(pt);
}

Gg2.worldIntersect = function (rayStart, rayDir, maxDist, rayStepThreshold) {

    var pt = rayStart;
    var rayLen = 0.0;

    var dist;

    while ((dist = Gg2.worldGetDist(pt)) > rayStepThreshold) {
        rayLen += dist;
        if (rayLen > maxDist) {
            rayLen = Gg2.noIntersection;
            break;
        }

        pt = pt.add(rayDir.multiply(dist));
    }

    return rayLen;
}

Gg2.putpixel = function (x, y, color) {
    if (x < 0 || x >= Gg2.viewWidth || y < 0 || y >= Gg2.viewHeight) {
        return;
    }

    Gg2.pixels[y * Gg2.viewWidth + x] = color;
}

Gg2.start = function () {
    //while (true)
    Gg2.run();
}
var numFrames = 0;
var fps = 0;
var lastFpsTime = 0;
Gg2.run = function () {
    Gg2.frame_timetocalc = Date.now();

    Gg2.lastatime += Gg2.speed;


    var moveDir = new Vec(1.5, Gg2.zz, 0.8);
    var cameraPos = moveDir.multiply(Gg2.lastatime);
    var cameraLookAt = cameraPos.add(moveDir).add(new Vec(0 + Gg2.rightm, 0.8 - Gg2.zz * 500, 0.5 + Gg2.leftm));
    var cameraUp = new Vec(0 + Gg2.rightm / 3, 1.0, 0 + Gg2.leftm / 3);
    var cameraForward = cameraLookAt.subtract(cameraPos).normalise();
    var cameraRight = cameraForward.crossProduct(cameraUp);
    cameraUp = cameraRight.crossProduct(cameraForward);

    if (!(Gg2.wlightprgrs[0] || Gg2.wlightprgrs[1] || Gg2.wlightprgrs[2])) {
        Gg2.wlightprgrs[0] = true;
    }
    for (var ij = 0; ij < 3; ij++) {
        if (Gg2.wlightprgrs[ij]) {
            if (Gg2.wlightstate[ij]) {
                Gg2.wlight[ij] -= 0.05;
                if (Gg2.wlight[ij] < 0.1) {
                    Gg2.wlight[ij] = 0.05;
                    Gg2.wlightstate[ij] = false;
                    Gg2.wlightprgrs[ij] = false;
                    if (ij != 2) {
                        Gg2.wlightprgrs[ij + 1] = true;
                    }
                }
            } else {
                Gg2.wlight[ij] += 0.05;
                if (Gg2.wlight[ij] > 1) {
                    Gg2.wlight[ij] = 1;
                    Gg2.wlightstate[ij] = true;
                    Gg2.wlightprgrs[ij] = false;
                    if (ij != 2) {
                        Gg2.wlightprgrs[ij + 1] = true;
                    }
                }
            }
            break;
        }
    }
    if (!Gg2.disable_ani) {
        if (Gg2.inverval + 0.05 > Gg2.maxinverval) {
            Gg2.inverval = Gg2.maxinverval;
            Gg2.invert = true;
        }
        if (Gg2.invert) {
            Gg2.inverval -= 0.05;
            if (Gg2.currcolor[0] && Gg2.currcolor[1]) {
                Gg2.invervalt[0] -= 0.05;
                Gg2.invervalt[2] += 0.05;
            } else if (Gg2.currcolor[1] && Gg2.currcolor[2]) {
                Gg2.invervalt[1] -= 0.05;
                Gg2.invervalt[0] += 0.05;
            } else if (Gg2.currcolor[2] && Gg2.currcolor[0]) {
                Gg2.invervalt[2] -= 0.05;
                Gg2.invervalt[1] += 0.05;
            }
        } else {
            Gg2.inverval += 0.05;
            if (Gg2.currcolor[0] && Gg2.currcolor[1]) {
                Gg2.invervalt[0] += 0.05;
                Gg2.invervalt[1] -= 0.05;
            } else if (Gg2.currcolor[1] && Gg2.currcolor[2]) {
                Gg2.invervalt[1] += 0.05;
                Gg2.invervalt[2] -= 0.05;
            } else if (Gg2.currcolor[2] && Gg2.currcolor[0]) {
                Gg2.invervalt[2] += 0.05;
                Gg2.invervalt[0] -= 0.05;
            }
        }
        if (Gg2.inverval - 0.04 > 1) {

        } else {
            Gg2.invert = false;
            Gg2.inverval = 1;

            if (Gg2.currcolor[0] && Gg2.currcolor[1]) {
                Gg2.inverval = Gg2.maxinverval;
                Gg2.currcolor[0] = false;
                Gg2.currcolor[2] = true;
            } else if (Gg2.currcolor[1] && Gg2.currcolor[2]) {
                Gg2.inverval = Gg2.maxinverval;
                Gg2.currcolor[1] = false;
                Gg2.currcolor[0] = true;
            } else if (Gg2.currcolor[2] && Gg2.currcolor[0]) {
                Gg2.inverval = Gg2.maxinverval;
                Gg2.currcolor[2] = false;
                Gg2.currcolor[1] = true;
            }
        }
    } else
        Gg2.inverval = Gg2.invervalt[0] = Gg2.invervalt[1] = Gg2.invervalt[2] = 1;

    for (var row = 0; row < Gg2.viewHeight; row++) {
        for (var col = 0; col < Gg2.viewWidth; col++) {

            var xFrac = (col - Gg2.viewWidth * 0.5) / Gg2.viewWidth;
            var yFrac = (row - Gg2.viewHeight * 0.5) / Gg2.viewWidth;
            var zFrac = 0.5;

            var rayStart = cameraPos;

            var rayDir = cameraRight.multiply(xFrac).add(cameraUp.multiply(yFrac)).add(cameraForward.multiply(zFrac));
            rayDir = rayDir.normalise();
            var depth = Gg2.worldIntersect(rayStart, rayDir, Gg2.maxViewDist, Gg2.primaryRayStepThreshold);

            if (depth == Gg2.noIntersection) {
                if (Gg2.scene == 2) {
                    Gg2.putpixel(col, row, Gg2.mat_yellow);
                } else {
                    Gg2.putpixel(col, row, Gg2.mat_blue);
                }
            } else {

                var dirToLight = new Vec(-1 - Gg2.wlight[1] * 2, 0.5 - Gg2.wlight[1] * 2, -0.7 - Gg2.wlight[2] * 2).normalise();
                var shadowFactor = Gg2.noIntersection;

                var pt = rayStart.add(rayDir.multiply(depth));
                var normal = Gg2.worldGetNormal(pt);


                if (shadowFactor == Gg2.noIntersection) {
                    shadowFactor = 1.0;
                } else {
                    shadowFactor = 1.0 - shadowFactor;
                }

                var normalisedIntensity = Math.min(Math.max(0.0, normal.dotProduct(dirToLight)) * shadowFactor, 1.0);
                var intensity = Math.round(normalisedIntensity * 255.0);

                var intensityr = Math.round(intensity * Gg2.invervalt[0]);
                var intensityg = Math.round(intensity * Gg2.invervalt[1]);
                var intensityb = Math.round(intensity * Gg2.invervalt[2]);
                if (Gg2.scene == 2) {
                    var depthByte = 0;
                    var invMaxViewDist = 1.0 / Gg2.maxViewDist;
                    if (depth != Gg2.noIntersection) {
                        depthByte = ((1.0 - depth * invMaxViewDist) * 255.0);
                        Gg2.putpixel(col, row, (Math.round(depthByte - intensity * 0.5) << 16) + (Math.round(depthByte - intensity * 0.5) << 8) + Math.round(depthByte));
                    }
                } else {

                    Gg2.putpixel(col, row, ((intensityr) << 16) + ((intensityg) << 8) + intensityb);

                }
            }
        }
    }

    Gg2.frame_timetocalc = Date.now() - Gg2.frame_timetocalc;
    if (Gg2.max_timetocalc < Gg2.frame_timetocalc) {
        Gg2.max_timetocalc = Gg2.frame_timetocalc;
    }
    if (Gg2.min_timetocalc > Gg2.frame_timetocalc) {
        Gg2.min_timetocalc = Gg2.frame_timetocalc;
    }
    numFrames++;
    var delta = (Date.now() - lastFpsTime) / Gg2.milsPerSecond;
    if (delta > Gg2.frameRateSmoothing) {
        fps = Math.round(numFrames / delta);
        numFrames = 0;
        lastFpsTime = Date.now();
    }
    //console.log("fps:" + fps + " , timetocalc:" + Gg2.frame_timetocalc + " ,  max:" + Gg2.max_timetocalc + " , min:" + Gg2.min_timetocalc);


}

Gg2.start();
